app/src/main/java/org/piwigo/data/repository/ImageRepository.java

```diff
@@ -327,30 +215,130 @@ public void updateImages(Integer categoryId) {
                         VariantWithImage vwi = new VariantWithImage();
                         vwi.image = i;
                         vwi.variant = new ImageVariant(i.id, i.width, i.height, location, null, d.url);
-                        Log.d("m_cache_sync","New variant "+vwi.image.name);
-                        return i;
+                        return new PositionedItem<VariantWithImage>(counter, vwi, true);
                     } else {
-                        Log.d("m_cache_sync","Existing variant "+existingVariant.image.name);
-                        return i;
+                        return new PositionedItem<>(counter, existingVariant, false);
                     }
                 });
-
-        restImages.subscribe();
-
-        Flowable<Image> dbImage = db.imageDao().getImagesInCategory(categoryId)
+        if(db == null) {
+            return remotes.toObservable();
+        }
+        return db.variantDao().getVariantsWithImageInCategory(categoryId)
             .subscribeOn(ioScheduler)
             .observeOn(ioScheduler)
             .flattenAsFlowable(s -> s)
+            .filter(variantWithImage -> {
+                if(!remoteIDs.contains(variantWithImage.image.id).blockingGet()) {
+                    Log.d("m_cache_sync","Deleted image "+variantWithImage.image.name);
```

Comment:
This filter operation makes the cache basically meaningless, as it seems to block the output until the remotes are fully there. Can we do it the other way round? And keep the local stream in addition. Then you could do the removal from the database in a separate thread in an onComplete callback for the remotes and filter there to remove the local variant.

In the first round I would accept, that the image is still shown (as we already emitted it) and only remove it from the cache to have it removed from the view only in the next refresh. In the long term we could emit at the very end - with another concatWith() the ones that are deleted - with the PositionedItem we should be able to emit a "removal" for a position (if we extend it). But this is in my eyes not critical for this release.